hera_pspec/uvwindow.py

```diff
@@ -0,0 +1,1320 @@
+        except ValueError:
+            raise ValueError("Wrong taper. See uvtools.dspec.gen_window"
+                             " for options.")
+        self.taper = taper
+
+        # polarisation
+        if isinstance(pols, str):
+            # if unique pol in string format (e.g. 'xx')
+            assert pols in ['pI', 'pQ', 'pV', 'pU', 'xx', 'yy', 'xy', 'yx'],\
+                "Wrong polarisation"
+            pols = (pols, pols)
+        elif isinstance(pols, int):
+            # if pol in integer format
+            if len(str(pols)) <= 2:
+                # if unique polarisation integer (e.g. -5)
+                assert pols in [1, 2, 4, 3, -5, -6, -7, -8], "Wrong polarisation"
+                # convert pol number to str according to AIPS Memo 117.
+                pols = (uvputils.polnum2str(pols, x_orientation=x_orientation),
+                        uvputils.polnum2str(pols, x_orientation=x_orientation))      
+            else:
+                # if polarisation pair integer
+                pols = uvputils.polpair_int2tuple(pols, pol_strings=True)
+        # if tuple of polarisation strings
+        elif isinstance(pols[0], str):
+            assert pols[0] and pols[1] in ['pI', 'pQ', 'pV', 'pU', 'xx', 'yy', 'xy', 'yx'],\
+                "Wrong polarisation"
+        # if tuple of polarisation integers
+        elif isinstance(pols[0], int):
+            assert pols[0] and pols[1] in [1, 2, 4, 3, -5, -6, -7, -8],\
+                "Wrong polarisation"
+            pols = (uvutils.polnum2str(pols[0], x_orientation=x_orientation),
+                    uvutils.polnum2str(pols[0], x_orientation=x_orientation))
+        else:
+            raise TypeError("Must feed pols as (tuple of) str or int.")
+        self.pols = pols
+
+        # spectral window
+        if spw_range is not None:
+            # checks on spw range input
+            spw_range = np.array(spw_range, dtype=int)
+            assert spw_range.size == 2, "spw_range must be fed as a tuple of "\
+                                        "frequency indices."
+            assert spw_range[1]-spw_range[0] > 0 and min(spw_range) >= 0,\
+                "Require non-zero spectral range."
+
+        # create list of FTBeam objects for each polarisation channel
+        self.ftbeam_obj_pol = []
+        # if not fed, creates appropriate FTBeam objects
+        if ftbeam_obj is None:
+            # create ftbeam_obj for first pol in the pair
+            self.ftbeam_obj_pol.append(FTBeam(self.pols[0], ftfile=ftfile, spw_range=spw_range,
+                                              verbose=verbose))
+            # copy object if polarisations in pair are identical
+            if self.pols[0] == self.pols[1]:
+                self.ftbeam_obj_pol.append(copy.deepcopy(self.ftbeam_obj_pol[0]))
+            # else, create new ftbeam object
+            else:
+                self.ftbeam_obj_pol.append(FTBeam(self.pols[1], ftfile=ftfile, spw_range=spw_range,
```

Comment:
I feel like it would be a LOT simpler to just accept a tuple of `FTBeam` objects into the constructor here. Then you let the `FTBeam` object deal with validating its own input. It's not any harder to call, really. If you really really want, you can have a constructor on `UVWindow` like the following:

```python
class UVWindow:
    @classmethod
    def from_ftfile(cls, ftfile, pols, **kwargs):
        ft = [FTBeam(ftfile, pol=pol, **kwargs) for pol in pols]
       return cls(ft_beam=ft)
```

Reply:
I am not sure I understand. The idea here is to be able to parse the FTBeam directly to UVWindow so that, if you've already computed it, you don't have to do it again. Since I've now changed FTBeam to have two class methods `from_file` and `from beam` I think it's made what you say impossible because I need to know before initialising the `FTBeam` object if it going to be initialized with a ft_file or not

Reply:
Okay, I think I got your point in the end. I've removed the option for `UVWindow` to be initialized with a ftfile object. It will always have to be fed one (or a tuple of) FTBeam object(s), it simplifies the code **a lot**

Reply:
I'm not sure I understand the difficulty here (probably my own thickness...). I envisage a workflow like the following:

```python
ftbeam_x = FTBeam.from_file(ftfile, pol='x')
ftbeam_y = FTBeam.from_file(ftfile, pol='x')

uvw = UVWindow(ftbeams=(ftbeam_x, ftbeam_y))
```

Since the `ftbeam_x` object has the `pol`, `spw` etc. defined on it already, you don't then need to pass that to `UVWindow`, you just need to verify that the two beams passed in are compatible. Your `from_uvpspec` function I presume would essentially just implement the above workflow but getting the beams from the UVPSpec object.

Am I missing something important here?


```

Reply:
Oops sorry didn't see your last response :-)

Reply:
No, I think I'm just trying to make the code more complicated than in needs to be (e.g. too many options for the user). Initializing the `UVWindow` attributes from the `FTBeam` ones seems like a good option